internal/expr/query.go

```diff
@@ -36,93 +36,142 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
 		}
 	}()
 
-	var typeInQuery = make(map[reflect.Type]bool)
-	for _, in := range pe.inputs {
-		typeInQuery[in.structType] = true
+	var inQuery = make(map[reflect.Type]bool)
+	for _, typeElement := range pe.inputs {
+		switch te := typeElement.(type) {
+		case field:
+			inQuery[te.structType] = true
+		}
 	}
 
 	var typeValue = make(map[reflect.Type]reflect.Value)
 	var typeNames []string
+	var m map[string]any
 	for _, arg := range args {
 		if arg == nil {
-			return nil, fmt.Errorf("need valid struct, got nil")
+			return nil, fmt.Errorf("need a map or struct, got nil")
 		}
 		v := reflect.ValueOf(arg)
 		v = reflect.Indirect(v)
 		t := v.Type()
 
-		if t.Kind() != reflect.Struct {
-			return nil, fmt.Errorf("need struct, got %s", t.Kind())
-		}
-
-		typeValue[t] = v
-		typeNames = append(typeNames, t.Name())
-
-		if !typeInQuery[t] {
-			// Check if we have a type with the same name from a different package.
-			for _, in := range pe.inputs {
-				if t.Name() == in.structType.Name() {
-					return nil, fmt.Errorf("type %s not found, have %s", in.structType.String(), t.String())
+		switch t.Kind() {
+		case reflect.Map:
+			if m != nil {
+				return nil, fmt.Errorf(`multiple maps`)
+			}
+			switch mtype := arg.(type) {
+			case map[string]any:
+				m = mtype
+			case *map[string]any:
```

Comment:
I'm not sure I follow, why are both the `map[string]any` and `*map[string]any` cases required? Would it be cleaner to enforce just one to be consistent with `*struct{}` being forbidden elsewhere? It also feels a little strange to see the pointer case guarded with a `.Kind() == reflect.Map`

Reply:
Ah okay thanks for the explanation. Consistency sounds good